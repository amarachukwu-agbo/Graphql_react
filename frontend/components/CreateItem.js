import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Form from './styles/Form';
import Error from './ErrorMessage';
import Router from 'next/router';

export const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String) {
    createItem(
      title: $title
      description: $description
      price: $price
      image: $image
      largeImage: $largeImage
    ) {
      id
    }
  }
`;
export default class CreateItem extends Component {
  state = {
    title: '',
    description: '',
    image: '',
    largeImage: '',
    price: 0
  }

  handleChange = (e) => {
    const { name, value, type } = e.target;
    const val = type === 'number' ? parseFloat(value) : value;
    this.setState({
      [name]: val
    });
  }

  uploadImage = async (e) => {
    const { files } = e.target;
    const data = new FormData();
    data.append('file', files[0]);
    data.append('upload_preset', 'sickfits');
    const res = await fetch('https://api.cloudinary.com/v1_1/ama-hello-books-v2/image/upload', {
      method: 'POST',
      mode: 'cors',
      body: data
    });
    const file = await res.json();
    this.setState({
      image: file.secure_url,
      largeImage: file.eager[0].secure_url
    });
  }
  render() {
    const { title, price, description, image } = this.state;
    return (
      <Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
      {(createItem, { loading, error}) => (
        <Form onSubmit={async (e) => {
          e.preventDefault();
          const res = await createItem();
          Router.push({
            pathname: '/item',
            query: { id: res.data.createItem.id}
          });
        }}>
          <h2> Sell an item</h2>
          <Error error={error} />
          <fieldset disabled={loading} aria-busy={loading}>
          <label htmlFor="file">
            File
            </label>
            <input
              type="file"
              id="file"
              name="file"
              placeholder="Upload an image"
              required
              onChange={this.uploadImage}
            />
            {
              image && <img src={image} alt="uploaded image" width="50px"/>
            }
            <label htmlFor="title">
            Title
            </label>
            <input
              type="text"
              id="title"
              name="title"
              placeholder="Title"
              required
              value={title}
              onChange={this.handleChange}
            />

            <label htmlFor="price">
            Price
            </label>
            <input
              type="number"
              id="price"
              name="price"
              placeholder="Price"
              required
              value={price}
              onChange={this.handleChange}
            />

            <label htmlFor="description">
            Description
            </label>
            <textarea
              type="text"
              id="description"
              name="description"
              placeholder="Enter a description"
              required
              value={description}
              onChange={this.handleChange}
            />
            <button type="submit">Submit</button>
          </fieldset>
        </Form>
      )}
      </Mutation>
    )
  }
}
