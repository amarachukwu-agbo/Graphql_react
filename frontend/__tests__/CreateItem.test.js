import { mount } from 'enzyme';
import wait from 'waait';
import Router from 'next/router';
import { MockedProvider } from 'react-apollo/test-utils';
import CreateItem, { CREATE_ITEM_MUTATION } from '../components/CreateItem';
import { fakeItem } from '../lib/testUtils';

Router.router = {
  push: jest.fn(),
}

const dogImage = 'dog-small.jpg';

global.fetch = jest.fn().mockResolvedValue({
  json: () => ({
    secure_url: dogImage,
    eager: [{ secure_url: dogImage }]
  })
})
const fakeCreateItem = fakeItem();

const mocks = [
  {
    request: {
      query: CREATE_ITEM_MUTATION,
      variables: {
        title: fakeCreateItem.title,
        description: fakeCreateItem.description,
        image: fakeCreateItem.image,
        largeImage: fakeCreateItem.largeImage,
        price: fakeCreateItem.price,
      }
    },
    result: {
      data: {
        createItem: {
          ...fakeCreateItem,
        }
      }
    }
  }
];

describe('<CreateItem />', () => {
  it('renders correctly and matches snapshot', async () => {
    const wrapper = mount(<MockedProvider mocks={mocks}>
      <CreateItem />
    </MockedProvider>);
    expect(wrapper.find('Form')).toMatchSnapshot();
  });

  it('uploads file input', async () => {
    const wrapper = mount(<MockedProvider mocks={mocks}>
      <CreateItem />
    </MockedProvider>);
    wrapper.find('input[type="file"]').simulate('change', {
      target: {
        files: ['fakeDog.jpg']
      }
    });
    await wait(0);
    const createItemState = wrapper.find('CreateItem').instance().state;
    expect(createItemState.image).toBe(dogImage);
    expect(createItemState.largeImage).toBe(dogImage);
    expect(global.fetch).toHaveBeenCalled();
  });

  it('handles state updating', () => {
    const wrapper = mount(<MockedProvider mocks={mocks}>
      <CreateItem />
    </MockedProvider>);
    wrapper.find('input[name="title"]').simulate('change', {
      target: {
        value: fakeCreateItem.title,
        name: "title"
      }
    });
    wrapper.find('input[name="price"]').simulate('change', {
      target: {
        value: fakeCreateItem.price,
        name: "price"
      }
    });
    wrapper.find('textarea[name="description"]').simulate('change', {
      target: {
        value: fakeCreateItem.description,
        name: "description"
      }
    });
    const createItemState = wrapper.find('CreateItem').instance().state;
    expect(createItemState).toMatchObject({
      title: fakeCreateItem.title,
      price: fakeCreateItem.price,
      description: fakeCreateItem.description,
    });
  });

  it('runs the mutation when form is submitted', async () => {
    const wrapper = mount(<MockedProvider mocks={mocks}>
      <CreateItem />
    </MockedProvider>);
    wrapper.find('input[type="file"]').simulate('change', {
      target: {
        files: ['fakeDog.jpg']
      }
    });
    wrapper.find('input[name="title"]').simulate('change', {
      target: {
        value: fakeCreateItem.title,
        name: "title"
      }
    });
    wrapper.find('input[name="price"]').simulate('change', {
      target: {
        value: fakeCreateItem.price,
        name: "price",
        type: "number"
      }
    });
    wrapper.find('textarea[name="description"]').simulate('change', {
      target: {
        value: fakeCreateItem.description,
        name: "description"
      }
    });
    await wait(0);
    wrapper.find('form').simulate('submit');
    await wait(50);
    expect(Router.router.push).toHaveBeenCalled();
    expect(Router.router.push)
      .toHaveBeenCalledWith({"pathname": "/item", "query": {"id": "abc123"}});
    global.fetch.mockReset();
  });
});
