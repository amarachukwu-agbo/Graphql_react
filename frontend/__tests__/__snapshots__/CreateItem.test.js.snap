// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateItem /> renders correctly and matches snapshot 1`] = `
<Form
  onSubmit={[Function]}
>
  <form
    className="Form-sc-1xszr8q-0 krRxky"
    onSubmit={[Function]}
  >
    <h2>
       Sell an item
    </h2>
    <DisplayError
      error={Object {}}
    />
    <fieldset
      aria-busy={false}
      disabled={false}
    >
      <label
        htmlFor="file"
      >
        File
      </label>
      <input
        id="file"
        name="file"
        onChange={[Function]}
        placeholder="Upload an image"
        required={true}
        type="file"
      />
      <label
        htmlFor="title"
      >
        Title
      </label>
      <input
        id="title"
        name="title"
        onChange={[Function]}
        placeholder="Title"
        required={true}
        type="text"
        value=""
      />
      <label
        htmlFor="price"
      >
        Price
      </label>
      <input
        id="price"
        name="price"
        onChange={[Function]}
        placeholder="Price"
        required={true}
        type="number"
        value={0}
      />
      <label
        htmlFor="description"
      >
        Description
      </label>
      <textarea
        id="description"
        name="description"
        onChange={[Function]}
        placeholder="Enter a description"
        required={true}
        type="text"
        value=""
      />
      <button
        type="submit"
      >
        Submit
      </button>
    </fieldset>
  </form>
</Form>
`;
